[docs] Create examples of pickers with custom fields (#8034)
mui/mui-x: docs/data/date-pickers/custom-field/custom-field.md

---

# Custom field

<p class="description">The Date and Time Pickers let you pass custom fields to your pickers</p>

## Customize the default field

### Customize the `TextField`

You can use the `textField` slot to pass custom props to the `TextField`:

{{"demo": "TextFieldSlotProps.js"}}

### Customize the separator of multi input fields [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)

You can use the `fieldSeparator` slot to pass custom props to the `Typography` rendered between the two `TextField`:

{{"demo": "MultiInputFieldSeparatorSlotProps.js"}}

## Commonly used custom field

### Using another input

#### With the Joy input

:::info
This demo demonstrates how to take advantage of the field smart behaviors with any input (even third party ones).
For Joy, we will provide a higher level solution in the coming months for an even simpler usage.
:::

{{"demo": "PickerWithJoyField.js", "defaultCodeOpen": false}}

#### With the browser input

You can also use any other input:

{{"demo": "PickerWithBrowserField.js", "defaultCodeOpen": false}}

### Using an `Autocomplete`

If your user can only select a value in a small list of available dates,

> you can replace the field by an `Autocomplete` which just lists the those dates:

Reviewer: ```suggestion
you can replace the field with an `Autocomplete` listsing those dates:
```

{{"demo": "PickerWithAutocompleteField.js", "defaultCodeOpen": false}}

### Using a `Button`

If you only want to allow the user to pick a value through the views,

> you can replace the field by a `Button`:

Reviewer: ```suggestion
you can replace the field with a `Button`:
```

{{"demo": "PickerWithButtonField.js", "defaultCodeOpen": false}}

## How to build a custom field

The main challenge when building a custom field, is to make sure that all the relevant props passed by the pickers are correctly handled.

On the examples below, you can see that the typing of the props received by a custom field always have the same shape:

```tsx
interface JoyDateFieldProps
  extends UseDateFieldProps<Dayjs>, // The headless field props
    BaseSingleInputFieldProps<Dayjs | null, DateTimeValidationError> {} // The DOM field props

interface JoyDateTimeFieldProps
  extends UseDateTimeFieldProps<Dayjs>, // The headless field props
    BaseSingleInputFieldProps<Dayjs | null, DateTimeValidationError> {} // The DOM field props
```

### The headless field props

> This interface depend on which type of field you are building (`UseDateField` for date field, `UseTimeField` for a time field, `UseDateRangeFieldProps` for a date range field, ...).

Reviewer: ```suggestion
This interface depends on which type of field you are building (`UseDateField` for a date field, `UseTimeField` for a time field, `UseDateRangeFieldProps` for a date range field, etc.).
```

It contains:

> - the basic props common to all the fields (`value`, `onChange`, `format`, `readOnly`, ...)

Reviewer: ```suggestion
- the basic props common to all the fields (`value`, `onChange`, `format`, `readOnly`, etc.)
```

- the validation props for this type of field (`minDate`, `maxDate`, `shouldDisableDate`, ...)

:::info

> If you are building a custom field that don't have any input editing (e.g: the _Button field_), you can ignore most of those props.

Reviewer: ```suggestion
If you are building a custom field that doesn't have any input editing (e.g: the _Button field_), you can ignore most of those props.
```

:::

### The DOM field props

> This interface contains the props the pickers passes to its field in order to customize the rendering.

Reviewer: ```suggestion
This interface contains the props the pickers pass to its field in order to customize the rendering.
```

> These props are shaped to be received by our built-in fields which are using the `TextField` from `@mui/material`.

Reviewer: ```suggestion
These props are shaped to be received by the built-in fields which are using the `TextField` from `@mui/material`.
```

> When used with another type of input (or no input at all), you will have to manually re-attribute them to the relevant component.

Reviewer: ```suggestion
When used with another type of input (or no input at all), you will have to manually pass them to the relevant component.
```

---
product: date-pickers
title: Date and Time pickers - Custom field
githubLabel: 'component: pickers'
packageName: '@mui/x-date-pickers'
---

# Custom field

<p class="description">The Date and Time Pickers let you customize the field by passing props or custom components</p>

## Customize the default field

### Customize the `TextField`

You can use the `textField` slot to pass custom props to the `TextField`:

{{"demo": "TextFieldSlotProps.js"}}

### Customize the separator of multi input fields [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)

You can use the `fieldSeparator` slot to pass custom props to the `Typography` rendered between the two `TextField`:

{{"demo": "MultiInputFieldSeparatorSlotProps.js"}}

## Commonly used custom field

> ### Using another input

Author: Can we do a follow up to properly introduce the various hooks and how the work without delaying this one for too long ?
